webcompat/static/js/lib/bugform.js

```diff
@@ -573,6 +599,179 @@ function BugForm() {
     });
   };
 
+  // functions for Image Editor
+  this.openImageEditor = function() {
+    var canvas = document.createElement('canvas');
+    var ctx = canvas.getContext('2d');
+    var modalContent = document.querySelector('.modal-content');
+
+    var img = new Image();
+    // grab the data URI from background-image property, since it's already
+    // in the DOM: 'url("data:image/....")'
+    img.src = this.previewEl.get(0).style.backgroundImage.slice(5, -2);
+    this.editImageModal.removeClass("is-hidden");
+
+    ctx.imageSmoothingEnabled = false;
+    ctx.webkitImageSmoothingEnabled = false;
+    ctx.msImageSmoothingEnabled = false;
+    ctx.imageSmoothingEnabled = false;
+
+    img.onload = _.bind(function() {
+      // used to resize screenshot for Image Editor
+      this.curScreenshotWidth = img.width || null;
+      this.curScreenshotHeight = img.height || null;
+
+      var canvasDimensions = this.calcCanvasDimensions();
+      canvas.width = canvasDimensions.width;
+      canvas.height = canvasDimensions.height;
+      modalContent.appendChild(canvas);
+      if (this.curScreenshotWidth && this.curScreenshotWidth < canvas.width) {
+        ctx.drawImage(img, 0, 0);
+      } else {
+        ctx.drawImage(img, 0, 0, canvas.width, canvas.height);
+      }
+
+      var domCanvas = $("canvas");
+      domCanvas.on("mousedown", {ctx: ctx}, _.bind(this.canvasMouseDown, this));
+      domCanvas.on("mousemove", {ctx: ctx}, _.bind(this.canvasMouseDrag, this));
+      domCanvas.on("mouseup", {ctx: ctx}, _.bind(this.canvasMouseUp, this));
+    }, this);
+  }
+
+  this.closeImageEditor = function() {
+    var canvas = document.querySelector('canvas');
+ 
+    this.dataURI = canvas.toDataURL();
+    this.editImageModal.addClass("is-hidden");
+    this.addPreviewBackground(this.dataURI);
+    canvas.remove();
+  }
+
+  this.setImageEditorHighlight = function() {
+    var imageEditorHighlightBtn = $(".js-highlight-button");
+    var imageEditorRedactBtn = $(".js-redact-button");
+
+    imageEditorHighlightBtn.toggleClass("is-active");
+    imageEditorRedactBtn.toggleClass("is-active");
+    this.curEditorActiveControl = "highlight";
+  }
+
+  this.setImageEditorRedact = function() {
+    var imageEditorHighlightBtn = $(".js-highlight-button");
+    var imageEditorRedactBtn = $(".js-redact-button");
+
+    imageEditorRedactBtn.toggleClass("is-active");
+    imageEditorHighlightBtn.toggleClass("is-active");
+    this.curEditorActiveControl = "redact";
+  }
+
+  this.canvasMouseDown = function(ev) {
+    var canvas = ev.target;
+    var mPos = this.getMousePos(canvas, ev);
+    var canvasWidth = canvas.offsetWidth + 1;
+    var canvasHeight = canvas.offsetHeight + 1;
+
+    this.initMouseDownPos = mPos;
+    this.mouseDown = true;
+    // used to store image version before shapes drawn during drag events
+    this.canvasImageData = ev.data.ctx.getImageData(0, 0, canvasWidth, canvasHeight);
```

Comment:
@miketaylr this seems like a good way to get a snapshot of the image before all the on drag shape drawing occurs. should I be concerned with the size of the data and memory for it?